pkg/ttnpb/field_mask_validation.go

```diff
@@ -0,0 +1,577 @@
+	"version_ids.firmware_version",
+	"version_ids.hardware_version",
+	"version_ids.model_id",
+}
+
+var isEndDeviceWriteFieldPaths = []string{
+	"application_server_address",
+	"attributes",
+	"description",
+	"ids",
+	"ids.dev_eui",
+	"ids.join_eui",
+	"join_server_address",
+	"locations",
+	"name",
+	"network_server_address",
+	"service_profile_id",
+	"version_ids",
+	"version_ids.brand_id",
+	"version_ids.firmware_version",
+	"version_ids.hardware_version",
+	"version_ids.model_id",
+}
+
+var allowedFieldMaskPathsForRPC = map[string][]string{
+	// Applications:
+	"/ttn.lorawan.v3.ApplicationRegistry/Get":                 ApplicationFieldPathsNested,
+	"/ttn.lorawan.v3.ApplicationRegistry/List":                ApplicationFieldPathsNested,
+	"/ttn.lorawan.v3.ApplicationRegistry/Update":              ApplicationFieldPathsNested,
+	"/ttn.lorawan.v3.EntityRegistrySearch/SearchApplications": ApplicationFieldPathsNested,
+
+	// Application Webhooks:
+	"/ttn.lorawan.v3.ApplicationWebhookRegistry/Get":  ApplicationWebhookFieldPathsNested,
+	"/ttn.lorawan.v3.ApplicationWebhookRegistry/List": ApplicationWebhookFieldPathsNested,
+	"/ttn.lorawan.v3.ApplicationWebhookRegistry/Set":  ApplicationWebhookFieldPathsNested,
+
+	// Application Links:
+	"/ttn.lorawan.v3.As/GetLink": ApplicationLinkFieldPathsNested,
+	"/ttn.lorawan.v3.As/SetLink": ApplicationLinkFieldPathsNested,
+
+	// Clients:
+	"/ttn.lorawan.v3.ClientRegistry/Get":                 ClientFieldPathsNested,
+	"/ttn.lorawan.v3.ClientRegistry/List":                ClientFieldPathsNested,
+	"/ttn.lorawan.v3.ClientRegistry/Update":              ClientFieldPathsNested,
+	"/ttn.lorawan.v3.EntityRegistrySearch/SearchClients": ClientFieldPathsNested,
+
+	// End Devices:
+	"/ttn.lorawan.v3.AsEndDeviceRegistry/Get": {
+		"formatters",
+		"formatters.down_formatter",
+		"formatters.down_formatter_parameter",
+		"formatters.up_formatter",
+		"formatters.up_formatter_parameter",
+		"ids",
+		"ids.application_ids",
+		"ids.application_ids.application_id",
+		"ids.dev_eui",
+		"ids.device_id",
+		"ids.join_eui",
+		"pending_session",
+		"pending_session.dev_addr",
+		"pending_session.keys.app_s_key",
+		"pending_session.keys.app_s_key.key",
+		"pending_session.keys.app_s_key.kek_label",
+		"pending_session.last_a_f_cnt_down",
+		"session",
+		"session.dev_addr",
+		"session.keys.app_s_key",
+		"session.keys.app_s_key.key",
+		"session.keys.app_s_key.kek_label",
+		"session.last_a_f_cnt_down",
+		"version_ids",
+		"version_ids.brand_id",
+		"version_ids.firmware_version",
+		"version_ids.hardware_version",
+		"version_ids.model_id",
+	},
+	"/ttn.lorawan.v3.AsEndDeviceRegistry/Set": {
+		"formatters",
+		"formatters.down_formatter",
+		"formatters.down_formatter_parameter",
+		"formatters.up_formatter",
+		"formatters.up_formatter_parameter",
+		"ids",
+		"ids.application_ids",
+		"ids.application_ids.application_id",
+		"ids.dev_eui",
+		"ids.device_id",
+		"ids.join_eui",
+		"pending_session",
```

Comment:
1. Is it safe to allow users to set this?
2. This allows setting any field in `pending_session`, e.g. `pending_session.keys.s_nwk_s_int_key` - does AS ensure this does not happen?
3. This allows deleting `pending_session`

I propose removing this field and be more restrictive(using the nested paths below)

Reply:
1. We need low-level access for device migration
2. Will change this to specific paths
3. See 1